In [128]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm

In [129]:
df = pd.read_csv('logs_notifications.csv')
df.head()

browser             datetime                     ip_anonymized language  \
0  firefox  2015-09-01 00:00:00  d588180c3e5c462692c24cffd961de43       ru   
1   chrome  2015-09-01 00:00:00  330f85e852916008aa80feb339696b2e       cn   
2   chrome  2015-09-01 00:00:00  b85871b1112cde31be2864927c035ae3       fr   
3   chrome  2015-09-01 00:00:00  51178bbf1c09a9a709b719fe948b5bd2       en   
4  firefox  2015-09-01 00:00:01  884eec370b6d6bdf20d625039b9a3e8b       en   

   message  
0        2  
1        0  
2        0  
3        2  
4        1

In [130]:
df.shape

(500000, 5)

In [131]:
df1 = pd.read_csv('logs_web.csv')
df1.head()

browser             datetime                     ip_anonymized language  \
0   chrome  2015-09-01 00:02:14  9631420eea21b0458d01f7cc598c3ac2       en   
1  firefox  2015-09-01 00:04:25  08bd4c63e1abb0e624cde4ac1518de73       ru   
2  firefox  2015-09-01 00:07:34  07678caba7be8bbadead98d5b530c5ff       de   
3   chrome  2015-09-01 00:07:47  ba759e7216a6a73d55591aed389ce2ad       en   
4   chrome  2015-09-01 00:08:41  db166a3780720484f9ca31b093c5b6f0       en   

   message  
0        0  
1        2  
2        1  
3        2  
4        1

In [132]:
df2.shape

(6438, 5)

In [133]:
df_new= df.merge(df1, on='ip_anonymized', how='left')
df_new


browser_x           datetime_x                     ip_anonymized  \
0        firefox  2015-09-01 00:00:00  d588180c3e5c462692c24cffd961de43   
1         chrome  2015-09-01 00:00:00  330f85e852916008aa80feb339696b2e   
2         chrome  2015-09-01 00:00:00  b85871b1112cde31be2864927c035ae3   
3         chrome  2015-09-01 00:00:00  51178bbf1c09a9a709b719fe948b5bd2   
4        firefox  2015-09-01 00:00:01  884eec370b6d6bdf20d625039b9a3e8b   
...          ...                  ...                               ...   
499995    chrome  2015-09-07 23:59:47  610e9adf09f64640e5715b3cdc5dc651   
499996    chrome  2015-09-07 23:59:49  62143d060d661e4c4032572207af3007   
499997    chrome  2015-09-07 23:59:50  3c84701f20eeadf652e9cb9ee25677eb   
499998    chrome  2015-09-07 23:59:53  0e27bc5703ed5d0fd73f22c790368449   
499999    chrome  2015-09-07 23:59:55  d117ac82c57cfe5c9cb422b275ac063a   

       language_x  message_x browser_y datetime_y language_y  message_y  
0              ru          2       NaN        NaN        NaN        NaN  
1              cn          0       NaN        NaN        NaN        NaN  
2              fr          0       NaN        NaN        NaN        NaN  
3              en          2       NaN        NaN        NaN        NaN  
4              en          1       NaN        NaN        NaN        NaN  
...           ...        ...       ...        ...        ...        ...  
499995         fr          0       NaN        NaN        NaN        NaN  
499996         fr          0       NaN        NaN        NaN        NaN  
499997         en          1       NaN        NaN        NaN        NaN  
499998         en          2       NaN        NaN        NaN        NaN  
499999         de          2       NaN        NaN        NaN        NaN  

[500000 rows x 9 columns]

In [134]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500000 entries, 0 to 499999
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   browser_x      500000 non-null  object 
 1   datetime_x     500000 non-null  object 
 2   ip_anonymized  500000 non-null  object 
 3   language_x     500000 non-null  object 
 4   message_x      500000 non-null  int64  
 5   browser_y      6384 non-null    object 
 6   datetime_y     6384 non-null    object 
 7   language_y     6384 non-null    object 
 8   message_y      6384 non-null    float64
dtypes: float64(1), int64(1), object(7)
memory usage: 38.1+ MB


In [135]:
df_new['Yes/No'] = np.where(df_new['message_y'].isnull(), 0, 1)
df_new

browser_x           datetime_x                     ip_anonymized  \
0        firefox  2015-09-01 00:00:00  d588180c3e5c462692c24cffd961de43   
1         chrome  2015-09-01 00:00:00  330f85e852916008aa80feb339696b2e   
2         chrome  2015-09-01 00:00:00  b85871b1112cde31be2864927c035ae3   
3         chrome  2015-09-01 00:00:00  51178bbf1c09a9a709b719fe948b5bd2   
4        firefox  2015-09-01 00:00:01  884eec370b6d6bdf20d625039b9a3e8b   
...          ...                  ...                               ...   
499995    chrome  2015-09-07 23:59:47  610e9adf09f64640e5715b3cdc5dc651   
499996    chrome  2015-09-07 23:59:49  62143d060d661e4c4032572207af3007   
499997    chrome  2015-09-07 23:59:50  3c84701f20eeadf652e9cb9ee25677eb   
499998    chrome  2015-09-07 23:59:53  0e27bc5703ed5d0fd73f22c790368449   
499999    chrome  2015-09-07 23:59:55  d117ac82c57cfe5c9cb422b275ac063a   

       language_x  message_x browser_y datetime_y language_y  message_y  \
0              ru          2       NaN        NaN        NaN        NaN   
1              cn          0       NaN        NaN        NaN        NaN   
2              fr          0       NaN        NaN        NaN        NaN   
3              en          2       NaN        NaN        NaN        NaN   
4              en          1       NaN        NaN        NaN        NaN   
...           ...        ...       ...        ...        ...        ...   
499995         fr          0       NaN        NaN        NaN        NaN   
499996         fr          0       NaN        NaN        NaN        NaN   
499997         en          1       NaN        NaN        NaN        NaN   
499998         en          2       NaN        NaN        NaN        NaN   
499999         de          2       NaN        NaN        NaN        NaN   

        Yes/No  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
499995       0  
499996       0  
499997       0  
499998       0  
499999       0  

[500000 rows x 10 columns]

#### Assess the performances of messages 0, 1 and 2. Which one should we use if we send notifications to all users? 

In [136]:
Dummy_1 = pd.get_dummies(df_new['message_x'],drop_first=True)
Dummy_2 = pd.get_dummies(df_new['language_x'],drop_first = True)
Dummy_3 = pd.get_dummies(df_new['browser_x'],drop_first = True)

In [146]:
# Merging the three dummies created for 'message', 'language_x', 'brower_y' 
    
f_final = pd.concat([df_new, Dummy_1, Dummy_2, Dummy_3], axis=1, join='outer')
f_final

browser_x           datetime_x                     ip_anonymized  \
0        firefox  2015-09-01 00:00:00  d588180c3e5c462692c24cffd961de43   
1         chrome  2015-09-01 00:00:00  330f85e852916008aa80feb339696b2e   
2         chrome  2015-09-01 00:00:00  b85871b1112cde31be2864927c035ae3   
3         chrome  2015-09-01 00:00:00  51178bbf1c09a9a709b719fe948b5bd2   
4        firefox  2015-09-01 00:00:01  884eec370b6d6bdf20d625039b9a3e8b   
...          ...                  ...                               ...   
499995    chrome  2015-09-07 23:59:47  610e9adf09f64640e5715b3cdc5dc651   
499996    chrome  2015-09-07 23:59:49  62143d060d661e4c4032572207af3007   
499997    chrome  2015-09-07 23:59:50  3c84701f20eeadf652e9cb9ee25677eb   
499998    chrome  2015-09-07 23:59:53  0e27bc5703ed5d0fd73f22c790368449   
499999    chrome  2015-09-07 23:59:55  d117ac82c57cfe5c9cb422b275ac063a   

       language_x  message_x browser_y datetime_y language_y  message_y  \
0              ru          2       NaN        NaN        NaN        NaN   
1              cn          0       NaN        NaN        NaN        NaN   
2              fr          0       NaN        NaN        NaN        NaN   
3              en          2       NaN        NaN        NaN        NaN   
4              en          1       NaN        NaN        NaN        NaN   
...           ...        ...       ...        ...        ...        ...   
499995         fr          0       NaN        NaN        NaN        NaN   
499996         fr          0       NaN        NaN        NaN        NaN   
499997         en          1       NaN        NaN        NaN        NaN   
499998         en          2       NaN        NaN        NaN        NaN   
499999         de          2       NaN        NaN        NaN        NaN   

        Yes/No  1  2  de  en  es  fr  ru  firefox  
0            0  0  1   0   0   0   0   1        1  
1            0  0  0   0   0   0   0   0        0  
2            0  0  0   0   0   0   1   0        0  
3            0  0  1   0   1   0   0   0        0  
4            0  1  0   0   1   0   0   0        1  
...        ... .. ..  ..  ..  ..  ..  ..      ...  
499995       0  0  0   0   0   0   1   0        0  
499996       0  0  0   0   0   0   1   0        0  
499997       0  1  0   0   1   0   0   0        0  
499998       0  0  1   0   1   0   0   0        0  
499999       0  0  1   1   0   0   0   0        0  

[500000 rows x 18 columns]

In [147]:
#final_merge = df_new.merge(Dummy_1, left_index = True, right_index = True)
#final_merge

In [148]:
#final_merge = df_new.merge(Dummy_2, left_index = True, right_index = True)
#final_merge

In [149]:
#final_merge = df_new.merge(Dummy_3, left_index = True, right_index = True)
#final_merge

##### Performances of messages 0, 1 and 2. Which one should we use if we send notifications to all users? 

In [150]:
Y = f_final['Yes/No']
X = f_final[[1, 2]]
X = sm.add_constant(X)
model = sm.Logit(Y, X)
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.068291
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 Yes/No   No. Observations:               500000
Model:                          Logit   Df Residuals:                   499997
Method:                           MLE   Df Model:                            2
Date:                Mon, 16 Nov 2020   Pseudo R-squ.:                0.001078
Time:                        21:14:17   Log-Likelihood:                -34146.
converged:                       True   LL-Null:                       -34183.
Covariance Type:            nonrobust   LLR p-value:                 1.001e-16
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -4.5011      0.024   -191.178      0.000      -4.547      -4.455
1              0.2643      0.031      8.439      0.000       0.203       0.326
2              0.1772      0.032      5.549      0.000       0.115       0.240
==============================================================================
"""

##### Message 1 above has more performance rate, with a coefficient of 0.26, and it should be condisered while sending out notifications to users

In [ ]:
#sns.countplot(x = message_x', data = final_merge, order = [0, 2, 1])

#### Translation manager thinks that one or several of the translations are crappy and look like a 4-year old could have written that. Can you confirm that? 
##### Translations to be considered

In [152]:
Y = f_final['Yes/No']
X = f_final[['de', 'en', 'fr', 'ru']]
X = sm.add_constant(X)
model = sm.Logit(Y,X)
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.066902
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 Yes/No   No. Observations:               500000
Model:                          Logit   Df Residuals:                   499995
Method:                           MLE   Df Model:                            4
Date:                Mon, 16 Nov 2020   Pseudo R-squ.:                 0.02140
Time:                        21:15:40   Log-Likelihood:                -33451.
converged:                       True   LL-Null:                       -34183.
Covariance Type:            nonrobust   LLR p-value:                     0.000
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -5.1048      0.044   -115.160      0.000      -5.192      -5.018
de             1.0461      0.050     21.072      0.000       0.949       1.143
en             1.0208      0.048     21.106      0.000       0.926       1.116
fr             0.9881      0.057     17.467      0.000       0.877       1.099
ru            -0.4857      0.073     -6.644      0.000      -0.629      -0.342
==============================================================================
"""

###### From the translations above, Deutch has more performance rate with a coefficient of 1.04, followed by English with 1.02 coefficient.

In [ ]:
# sns.countplot(x = 'language', data = merged, order['en', 'de', 'fr', 'cn', 'ru'])

#### The designer is interested if there are differences between browsers, because the mechanisms for displaying notifications for these differ. 
###### The browsers to be considered based on their performance rate

In [153]:
Y = f_final['Yes/No']
X = f_final['firefox']
X = sm.add_constant(X)
model = sm.Logit(Y,X)
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.068060
         Iterations 9


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 Yes/No   No. Observations:               500000
Model:                          Logit   Df Residuals:                   499998
Method:                           MLE   Df Model:                            1
Date:                Mon, 16 Nov 2020   Pseudo R-squ.:                0.004458
Time:                        21:15:54   Log-Likelihood:                -34030.
converged:                       True   LL-Null:                       -34183.
Covariance Type:            nonrobust   LLR p-value:                 3.051e-68
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -4.5204      0.017   -269.386      0.000      -4.553      -4.488
firefox        0.4493      0.025     17.683      0.000       0.400       0.499
==============================================================================
"""